File: src/main/kotlin/com/elpassion/intelijidea/task/edit/MFBeforeRunTaskDialog.java
package com.elpassion.intelijidea.task.edit;

import com.elpassion.intelijidea.task.MFBeforeRunTask;
import com.elpassion.intelijidea.task.MFTaskData;
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.*;
import com.intellij.util.IconUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class MFBeforeRunTaskDialog extends DialogWrapper implements TaskEditForm {

    private final TaskEditValidator taskEditValidator = new TaskEditValidator(this);
    private final Project project;
    private LabeledComponent<TextFieldWithBrowseButton> mainframerToolHolder;
    public JPanel panel;
    public JTextField buildCommandField;
    public JTextField taskField;
    public TextFieldWithBrowseButton mainframerToolField;

    public MFBeforeRunTaskDialog(Project project) {
        super(project);
        this.project = project;
        setModal(true);
        init();
    }

    @Nullable
    @Override
    protected ValidationInfo doValidate() {
        return taskEditValidator.doValidate();
    }

    @Nullable
    @Override
    protected JComponent createCenterPanel() {
        return panel;
    }

    private void createUIComponents() {
        mainframerToolField = new TextFieldWithBrowseButton();
        mainframerToolField.setButtonIcon(IconUtil.getAddIcon());
        TextBrowseFolderListener textBrowseFolderListener = new TextBrowseFolderListener(FileChooserDescriptorFactory.createSingleFolderDescriptor(), project);
        mainframerToolField.addBrowseFolderListener(textBrowseFolderListener);
        mainframerToolHolder = new LabeledComponent<>();
        mainframerToolHolder.setComponent(mainframerToolField);
    }

    public void restoreMainframerTask(MFBeforeRunTask beforeRunTask) {
        MFTaskData data = beforeRunTask.getData();
        mainframerToolField.setText(data.getMainframerPath());
        buildCommandField.setText(data.getBuildCommand());
        taskField.setText(data.getTaskName());
    }

    public MFTaskData createMFTaskDataFromForms() {
        return new MFTaskData(mainframerToolField.getText(),
                buildCommandField.getText(),
                taskField.getText());
    }

    @NotNull
    @Override
    public JTextField taskField() {
        return taskField;
    }

    @NotNull
    @Override
    public JTextField buildCommandField() {
        return buildCommandField;
    }

    @NotNull
    @Override
    public TextFieldWithBrowseButton mainframerToolField() {
        return mainframerToolField;
    }
}

File: src/main/kotlin/com/elpassion/intelijidea/task/edit/MFBeforeRunTaskDialog.kt
package com.elpassion.intelijidea.task.edit

import com.elpassion.intelijidea.common.BuildCommandValidator
import com.elpassion.intelijidea.common.MainframerPathValidator
import com.elpassion.intelijidea.common.validateForm
import com.elpassion.intelijidea.task.MFTaskData
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import javax.swing.JComponent

class MFBeforeRunTaskDialog(project: Project) : DialogWrapper(project) {

    val form = MFBeforeRunTaskForm(project)

    init {
        isModal = true
        title = "Configure Mainframer Make"
        init()
    }

    override fun createCenterPanel(): JComponent = form.panel

    override fun doValidate() = with(form) {
        validateForm(MainframerPathValidator(mainframerToolField), BuildCommandValidator(buildCommandField))
    }

    fun restoreMainframerTaskData(data: MFTaskData) {
        form.mainframerToolField.text = data.mainframerPath
        form.buildCommandField.text = data.buildCommand
    }

    fun createMFTaskDataFromForms() = MFTaskData(
            buildCommand = form.buildCommandField.text,
            mainframerPath = form.mainframerToolField.text)
}